node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1720 @@
+		while let Some(msg) = self.rx_to.next().await {
+			match msg {
+				ToJob::CandidateBacking(msg) => {
+					self.process_msg(msg).await?;
+				}
+				_ => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn issue_candidate_invalid_message(
+		&mut self,
+		candidate: AbridgedCandidateReceipt,
+	) -> Result<(), Error> {
+		self.tx_from.send(FromJob::CandidateSelection(
+			CandidateSelectionMessage::Invalid(self.parent, candidate)
+		)).await?;
+
+		Ok(())
+	}
+
+	/// Validate the candidate that is requested to be `Second`ed and distribute validation result.
+	async fn validate_and_second(
+		&mut self,
+		candidate: AbridgedCandidateReceipt,
+		pov: PoVBlock,
+	) -> Result<ValidationResult, Error> {
+		let valid = self.request_candidate_validation(candidate.clone(), pov.clone()).await?;
+		let statement = match valid.0 {
+			ValidationResult::Valid => {
+				// make PoV available for later distribution. Send data to the availability
+				// store to keep. Sign and dispatch `valid` statement to network if we
+				// have not seconded the given candidate.
+				self.make_pov_available(pov, valid.1, valid.2).await?;
+				self.issued_validity.insert(candidate.hash());
+				Statement::Seconded(candidate)
+			}
+			ValidationResult::Invalid => {
+				let candidate_hash = candidate.hash();
+				self.issue_candidate_invalid_message(candidate).await?;
+				Statement::Invalid(candidate_hash)
+			}
+		};
+
+		if let Some(signed_statement) = self.sign_statement(statement) {
+			self.import_statement(&signed_statement).await?;
+			self.distribute_signed_statement(signed_statement).await?;
+		}
+
+		Ok(valid.0)
+	}
+
+	fn get_backed(&self) -> Vec<NewBackedCandidate> {
+		let proposed = self.table.proposed_candidates(&self.table_context);
+		let mut res = Vec::with_capacity(proposed.len());
+
+		for p in proposed.into_iter() {
+			let TableAttestedCandidate { candidate, validity_votes, .. } = p;
+
+			let (ids, validity_votes): (Vec<_>, Vec<_>) = validity_votes
+						.into_iter()
+						.map(|(id, vote)| (id, vote.into()))
+						.unzip();
+
+			let group = match self.table_context.groups.get(&self.assignment) {
+				Some(group) => group,
+				None => continue,
+			};
+
+			let mut validator_indices = BitVec::with_capacity(
+				group.len()
+			);
+
+			validator_indices.resize(group.len(), false);
+
+			for id in ids.iter() {
+				if let Some(position) = group.iter().position(|x| x == id) {
+					validator_indices.set(position, true);
+				}
+			}
+
+			let backed = BackedCandidate {
+				candidate,
+				validity_votes,
+				validator_indices,
+			};
+
+			res.push(NewBackedCandidate(backed.clone()));
+		}
+
+		res
+	}
+
+	/// Check if there have happened any new misbehaviors and issue necessary messages.
+	async fn issue_new_misbehaviors(&mut self) -> Result<(), Error> {
+		let mut reports = Vec::new();
+
+		for (k, v) in self.table.get_misbehavior().iter() {
+			if !self.reported_misbehavior_for.contains(k) {
+				self.reported_misbehavior_for.insert(*k);
```

Comment:
This seems to let a validator slide if we have already reported them for a difference offence. Slashing should be more eager than that

Reply:
ATM we configured these as 100% slashes but we will probably scale that back somewhat.

Reply:
@montekki file as a follow-on issue?

Reply:
yes right now we are a bit limited with the implementation of the `statement-table` which holds a 
```rust
	detected_misbehavior: HashMap<C::AuthorityId, MisbehaviorFor<C>>,
```

and what we get from it is this API:

```rust

	/// Access all witnessed misbehavior.
	pub fn get_misbehavior(&self)
		-> &HashMap<C::AuthorityId, MisbehaviorFor<C>>

```

so it looks like currently there is no way to get multiple offences out of it if I am not mistaken. I will file an issue for that.